### PR TITLE
Add option to set Newton-Raphson tolerance for interpolation

### PR DESCRIPTION
**Pull request by kpamnany**
### Description

Setting up a cubed sphere interpolated grid can fail when the DG grid is stretched beyond a certain point. Allowing control of the tolerance is a temporary fix until @skandalaCLIMA implements a proper one.

Also cleaned up the interface to `InterpolationConfiguration` a little.

- [X] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [X] Unit tests are included OR N/A.
- [X] Code is exercised in an integration test OR N/A.
- [X] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by kpamnany** [comment]
@szy21: you would use this by calling [this version](https://github.com/CliMA/ClimateMachine.jl/pull/1942/files#diff-93b9bcbb766d84e0b28e67035a67ad4b796f8079acd70774bfbf1401c417cd87R29-R34) of `InterpolationConfiguration`. Please test to see if this solves the problem you saw?

**Comment by kpamnany** [comment]
bors try

**Review by skandalaCLIMA — APPROVED** [review]

**Comment by kpamnany** [comment]
It isn't just `VariableTemplates` stuff @charleskawczynski. :-/

**Comment by szy21** [comment]
> @szy21: you would use this by calling [this version](https://github.com/CliMA/ClimateMachine.jl/blob/kp/interp-tol/src/Driver/diagnostics_configs.jl#L29-L34) of `InterpolationConfiguration`. Please test to see if this solves the problem you saw?

@jiahe23 Could you test this? Or you could push the driver using grid stretching and I can test it. Thanks!

**Comment by jiahe23** [comment]
I run this branch together with the `grid_stretching` feather enabled in `as/gcm-stretchgrid` and set `grid_stretching = SingleExponentialStretching(4.5)` in `ClimateMachine.AtmosGCMConfiguration(...)` in the `solid_body_rotation_fvm.jl` and got the following error

```
ERROR: LoadError: UndefVarError: grid not defined
Stacktrace:
 [1] InterpolationConfiguration(::StackedCubedSphereTopology{Float64,2}, ::ClimateMachine.DriverConfiguration{Float64}, ::Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}}; nr_toler::Nothing) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:35
 [2] InterpolationConfiguration(::StackedCubedSphereTopology{Float64,2}, ::ClimateMachine.DriverConfiguration{Float64}, ::Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}}) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:35
 [3] InterpolationConfiguration(::ClimateMachine.DriverConfiguration{Float64}, ::Array{Float64,2}, ::Tuple{Float64,Float64,Float64}; axes::Nothing) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:99
 [4] InterpolationConfiguration(::ClimateMachine.DriverConfiguration{Float64}, ::Array{Float64,2}, ::Tuple{Float64,Float64,Float64}) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:78
 [5] config_diagnostics(::Type{T} where T, ::ClimateMachine.DriverConfiguration{Float64}) at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:204
 [6] main() at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:159
 [7] top-level scope at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:220
 [8] include(::Function, ::Module, ::String) at ./Base.jl:380
 [9] include(::Module, ::String) at ./Base.jl:368
 [10] exec_options(::Base.JLOptions) at ./client.jl:296
 [11] _start() at ./client.jl:506
in expression starting at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:220
```

**Comment by jiahe23** [comment]
Since this PR is to fix the interpolation error occurred in the grid stretching diagnostics initialization. I suggest we first merge `as/gcm-strecthgrid`. @akshaysridhar

**Comment by jiahe23** [comment]
> I run this branch together with the `grid_stretching` feather enabled in `as/gcm-stretchgrid` and set `grid_stretching = SingleExponentialStretching(4.5)` in `ClimateMachine.AtmosGCMConfiguration(...)` in the `solid_body_rotation_fvm.jl` and got the following error
> 
> ```
> ERROR: LoadError: UndefVarError: grid not defined
> Stacktrace:
>  [1] InterpolationConfiguration(::StackedCubedSphereTopology{Float64,2}, ::ClimateMachine.DriverConfiguration{Float64}, ::Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}}; nr_toler::Nothing) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:35
>  [2] InterpolationConfiguration(::StackedCubedSphereTopology{Float64,2}, ::ClimateMachine.DriverConfiguration{Float64}, ::Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}}) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:35
>  [3] InterpolationConfiguration(::ClimateMachine.DriverConfiguration{Float64}, ::Array{Float64,2}, ::Tuple{Float64,Float64,Float64}; axes::Nothing) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:99
>  [4] InterpolationConfiguration(::ClimateMachine.DriverConfiguration{Float64}, ::Array{Float64,2}, ::Tuple{Float64,Float64,Float64}) at /home/jiahe/CliMA/ClimateMachine.jl/src/Driver/diagnostics_configs.jl:78
>  [5] config_diagnostics(::Type{T} where T, ::ClimateMachine.DriverConfiguration{Float64}) at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:204
>  [6] main() at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:159
>  [7] top-level scope at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:220
>  [8] include(::Function, ::Module, ::String) at ./Base.jl:380
>  [9] include(::Module, ::String) at ./Base.jl:368
>  [10] exec_options(::Base.JLOptions) at ./client.jl:296
>  [11] _start() at ./client.jl:506
> in expression starting at /home/jiahe/CliMA/ClimateMachine.jl/experiments/TestCase/solid_body_rotation_fvm.jl:220
> ```

Actually, this is not specific for the grid-sctretching set-up. I run this branch without grid-stretching and got the same error.

**Comment by kpamnany** [comment]
Sorry about that. It's fixed and I tested it with `solid_body_rotation_fvm.jl` from `as/config-stretchgrid`.

**Comment by kpamnany** [comment]
> Sorry about that. It's fixed and I tested it with `solid_body_rotation_fvm.jl` from `as/config-stretchgrid`.

Ah, now I see that this doesn't use the `InterpolationConfiguration` function that was added. So can you test it @jiahe23?

File: experiments/TestCase/baroclinic_wave.jl [hunk 1]
```diff
@@ -316,17 +316,10 @@ function config_diagnostics(FT, driver_config)
         step = FT(1000), # in m
     ))
 
-    #boundaries = [
-    #    FT(-90.0) FT(-180.0) _planet_radius
-    #    FT(90.0) FT(180.0) FT(_planet_radius + info.domain_height)
-    #]
-
-    #resolution = (FT(2), FT(2), FT(1000)) # in (deg, deg, m)
```
Comment (jiahe23) [inline experiments/TestCase/baroclinic_wave.jl:324]: These comments are kept intentionally. If one does not need the spectra diagnostics group,  then there is no need to compute the lat/long based on the Gaussian points.

File: src/Driver/diagnostics_configs.jl [hunk 2]
```diff
@@ -16,26 +16,67 @@ mutable struct DiagnosticsConfiguration
...
+`resolution` is specified, in which case the axes are set up with
+equi-distant points, or the `axes` may be specified directly (in
+lat/lon/lvl or x/y/z order).
 """
 function InterpolationConfiguration(
     driver_config::DriverConfiguration,
     boundaries::Array,
-    resolution::Tuple,
+    resolution = nothing;
+    axes = nothing,
```
Comment (szy21) [inline src/Driver/diagnostics_configs.jl:76]: Should we add the nr_toler keyword to this function as well?

Comment (kpamnany) [reply src/Driver/diagnostics_configs.jl:76]: We could. But, it doesn't apply to `InterpolationBrick`, and it's a temporary solution so I thought the simplest approach would be best.

**Comment by kpamnany** [comment]
bors try

**Comment by kpamnany** [comment]
The [change](https://github.com/CliMA/ClimateMachine.jl/pull/1942/files#diff-82b3975c329b9d66b0a61522d4f3619782472c4fc154d0d7d8fae6da9fa8cc13R84) to `solid_body_rotation_fvm.jl` have taken CI runtime from [5 minutes](https://buildkite.com/clima/climatemachine-ci/builds/1086#384ecf4a-c865-4c1c-8d88-c4e2ed44eef9) to over [1 hour](https://buildkite.com/clima/climatemachine-ci/builds/1095#cb4a5712-cfc0-4d8e-a541-22a2e0923aad). This slowdown has nothing to do with interpolation, since that only runs 3 times.

Is this much slowdown expected from grid stretching?

File: src/Driver/diagnostics_configs.jl (hunk 2 above)
Comment (szy21) [reply src/Driver/diagnostics_configs.jl:77]: ok, thanks! So is the change in solid_body_rotation_fvm.jl the right way to use the tolerance?

**Comment by szy21** [comment]
> The [change](https://github.com/CliMA/ClimateMachine.jl/pull/1942/files#diff-82b3975c329b9d66b0a61522d4f3619782472c4fc154d0d7d8fae6da9fa8cc13R84) to `solid_body_rotation_fvm.jl` have taken CI runtime from [5 minutes](https://buildkite.com/clima/climatemachine-ci/builds/1086#384ecf4a-c865-4c1c-8d88-c4e2ed44eef9) to over [1 hour](https://buildkite.com/clima/climatemachine-ci/builds/1095#cb4a5712-cfc0-4d8e-a541-22a2e0923aad). This slowdown has nothing to do with interpolation, since that only runs 3 times.
> 
> Is this much slowdown expected from grid stretching?

Yes, it's using explicit timestepping, and the minimal vertical nodal distance decrease by a factor of ~10 from grid stretching. I will change the exponential stretching and/or reduce timeend.

File: src/Driver/diagnostics_configs.jl (hunk 2 above)
Comment (kpamnany) [reply src/Driver/diagnostics_configs.jl:77]: Yes, that's what I had in mind. :)

**Review by szy21 — APPROVED** [review]

**Pull request merged**